### ISSUE TITLE
Variable length sequence for RNN?

### ISSUE THREAD

**Issue by RoastedPecans** [fchollet/keras#6776]
Hey everyone!

I'm trying to use a LSTM with data in the shape something like [300, 1, x] where x changes. As of right now I'm training the model where x = 300 but after training the model I'm trying to go back through and make predictions on data in the shape [300, 1, 270] for an example. I've been exploring this and it seems like it can be done with the Theano backend (see https://github.com/fchollet/keras/issues/2054) but in my code when I use the following model:

```
model = Sequential()
model.add(LSTM(50, input_shape=None))
model.add(Dense(1))
```

I get the error 'TypeError: 'NoneType' object is not iterable'. Anyone know what's going on? According to the issue I linked that should work fine unless I'm misinterpreting it (is it only the 1st dimension that can be None?).

Info:
Python v3.5.3
Keras v2.0.2
Theano v0.9.0

Thanks!

**Comment by bstriner** [comment]
`input_shape=(None,1)`. `input_shape` is a tuple. LSTM expects (batch x sequence x features). You need 3 dims even if the last dim is 1. First dim is assumed, so input_shape must have 2 entries for the LSTM. This will work for data like `[300,x,1]` where the x changes.

Cheers

**Comment by v1nc3nt27** [comment]
@bstriner Hello, could you please help me with a similar problem? I'm passing [batch_size, x, 100] into a LSTM: 

`LSTM(100, return_sequences=True, stateful=True, input_shape=(None, 100),
                 batch_input_shape=(200, None, 100))`

I'm fitting the model on the following randomly created matrices:

`        x_train = np.random.random((1000, 50, 100))`
`         x_train_2 = np.random.random((1000, 10,100))`

As far as I understood LSTMs (and the keras implementation) correctly, the x should refer to the number of LSTM cells. How is it possible to pass varying sequence lengths into them without padding up to a max. specified length? 
For each LSTM a state and three matrices have to be learned. It's even possible to pass another x_train_3 with a sequence lengths of 60 afterwards, but there shouldn't be states and matrices for the extra 10 cells. So I really don't know what's happening there.

By the way, I'm using Keras version 1.0.8 and Tensorflow GPU 0.9.

Here is my example code:

    from keras.models import Sequential
    from keras.layers import LSTM, Dense
    import numpy as np
    from keras import backend as K

    with K.get_session():

        # create model
        model = Sequential()
        model.add(LSTM(100, return_sequences=True, stateful=True, input_shape=(None, 100),
                 batch_input_shape=(200, None, 100)))
        model.add(LSTM(100))
        model.add(Dense(2, activation='softmax'))
        model.compile(loss='categorical_crossentropy',
                      optimizer='rmsprop',
                      metrics=['accuracy'])

        # Generate dummy training data
        x_train = np.random.random((1000, 50, 100))
        x_train_2 = np.random.random((1000, 10, 100))
        y_train = np.random.random((1000, 2))
        y_train_2 = np.random.random((1000, 2))

        # Generate dummy validation data
        x_val = np.random.random((200, 50, 100))
        y_val = np.random.random((200, 2))

        # fit and eval models
        model.fit(x_train, y_train, batch_size=200, nb_epoch=1, shuffle=False, validation_data=(x_val, y_val), verbose=1)
        model.fit(x_train_2, y_train_2, batch_size=200, nb_epoch=1, shuffle=False, validation_data=(x_val, y_val), verbose=1)
        score = model.evaluate(x_val, y_val, batch_size=200, verbose=1)`

**Comment by bstriner** [comment]
@v1nc3nt27 two different things to discuss.

1. Using the `None` means you can have a different batch size for each batch. Use `fit_generator` instead of `fit`. One batch can be `(1, 15, 100)`, the next batch can be `(1, 24, 100)`, whatever. That will mean you are only doing batches of size 1 so it will be noisy, but you can easily write something to have different lengths in each batch.

2. If you want different lengths within a single batch, you need to use masking. You can signal in-band or out-of-band. In-band would be a special value that indicates that the step should be ignored. Out-of-band would be a separate input with `1` for included steps and `0` for excluded steps. Multiply your loss by the mask, so losses for excluded steps are 0. 

If you do both of the above, each batch will be the length of the longest sequence in that batch. The shorter sequences in that batch will be padded and masked. This should be the most efficient method and the least amount of data each time.

It would be best to do both. If you use `fit`, everything needs to be padded out to the length of the longest sequence. If there are some really long outliers, that will have a huge impact.

Here is a somewhat complicated example but includes one way of doing the above. This experiment concatenates input and output, so only uses `max(input len+output len)` instead of `max(input)+max(output)`, but that is a proof-of-concept and not necessary.
https://github.com/bstriner/keras-seq2seq

Not the cleanest code, but if you understand how it works then you should be able to do anything you want.

Cheers

**Comment by v1nc3nt27** [comment]
@bstriner Thank you very much for your detailed answer. From the first look, the code seems absolutely readable, but still complex, since I'm new to this. I'll work though it.

Still, I think I got a problem in understanding LSTMs conceptually. If I understand correctly, method 1 means you don't need padding. Is this method possible, because after each batch the LSTM cells are being regenerated? That'd mean the states between sequences (= batches in case of batch_size=1) i and i+1 are not kept and the needed amount of LSTM cells (= sequence length)  is generated and randomly initialized for each batch again and again.
This would explain why my code above is working. The second time I call model.fit() on x_train_2, new LSTM cells in the needed size (and thus new matrices and states for the cells) would be randomly initialized. But this would also diminish the reason to use LSTMs in the first place. 
If the LSTM cells aren't being newly generated after each batch, where do the LSTM cells (and their states/matrices) come from, if a new batch has a sequence longer than in any other batch before? 

Greetings

**Comment by bstriner** [comment]
Method 1 and 2 aren't exclusive. If you just do 1, you don't need padding, because there is only one sequence in each batch.

States between different batches are not usually kept or relevant. `stateful` option in keras isn't something I would recommend unless you are sure about what you are doing. `stateful` retains the state between batches but doesn't backprop between batches. So it uses the previous state but it can't learn to make the previous state useful. 

Lot's of people run into issues with `stateful` thinking it does more than it actually does. All LSTMs are stateful, `stateful` is just for state between batches, which is normally a bad idea. Only useful if you have some insanely long sequences that you can't possibly fit into a single batch.

Learning only works within a batch. If you need to learn a dependency that is 20 steps away you should have at least 20 steps in each batch. You could do 2 stateful batches of 10 but it would be random whether it learns anything useful.

The simplest and best method for training LSTMs is to take many different subsequences and run them all through the LSTM (not `stateful`). That way you are trying to learn an LSTM that can pick up anywhere in the sequence and still do something useful.

The initial state of the LSTM is something that is learned. That initial state is the same for each sequence in a batch. It will learn the best initial state that will work across all of your examples.

tldr; don't use `stateful` unless you have a good reason. Training is more stable without it.

Machine learning is all about learning something that can generalize in a meaningful way. Best way to do that is to have many different subsequences and an LSTM that works for all of them. More ways you can slice and dice your data the better.

As an extreme example, imagine an LSTM trained on the entire brown corpus as one sequence in one batch. It would learn but it would be pointless. An LSTM that can predict variable length sub-sequences starting from anywhere in the corpus would be learning something meaningful.

Cheers

**Comment by bstriner** [comment]
@v1nc3nt27 Let me know your thoughts on what was confusing and how to make it less so. I think `stateful` has been causing a lot of issues.

Just submitted this issue for discussion: https://github.com/fchollet/keras/issues/7212

**Comment by v1nc3nt27** [comment]
Once more, thank you for your help.

I'll shortly describe what I actually want to do - this might clear up some things:
For my problem, I've got 5000 text documents with variable length. I embed each token of 
the documents in a 100d vector. I can't split the documents up into sequences (I want to keep it as simple as possible) and I only have a binary label for each doc. I guess the only reasonable way to go here, would be one big batch of 5000 and padding of shorter sequences, if I want to learn relations between the documents' sequences. So I'd have [5000, x, 100]. It might not be the usual task for which you use RNN's, but I want to test it and hope for some magic to happen. I thought `stateful` would make total sense in this case, but since you said it doesn't backprop between the states, I don't see how this would be helpful then.

ALso, after reading your explanation I wonder if using a batch size of 1 in connection with
LSTMs (without `stateful=true`) can be useful in any scenario. There'd be a single BPTT, but since the states get reset for the next batch, the changed values wouldn't effect the next batch and be useless.

In any case, what is still troubling me are the following two things:
1. If the states get reset after each batch, what happens at prediction time? Is the LSTM just using randomly initialized matrices? 
2. I doubt it, but anyway I want to ask: Are the LSTM states of each batch kept for the next epoch? So, are e.g. the LSTM states of batch no. 1 in epoch no. 1 being used for batch no. 1 in epoch no. 2?

Maybe renaming  `stateful` in LSTMs to something different and adding a short example to the documentation might help. Or at least explicitly stating, that `stateful=False` doesn't mean the LSTM cells do not save states anymore. I for myself was thinking, that using this argument is the only way to keep the state of an LSTM between sequences.

**Comment by fnozarian** [comment]
@v1nc3nt27 Did you find out the answers to your last two questions?

**Comment by th4t-gi** [comment]
I guess I'm a little late to the party but I'm was hoping for some answers to this issue and then... nothing. @bstriner, @v1nc3nt27 have you found any answers to v1nc3nt27's questions?

**Comment by yashgarg1232** [comment]
Any idea on how to configure validation_data with unequal length time series? I got the model to train on time series of unequal length using a batch size of 1.

**Comment by Deltaidiots** [comment]
> @v1nc3nt27 two different things to discuss.
> 
> 1. Using the `None` means you can have a different batch size for each batch. Use `fit_generator` instead of `fit`. One batch can be `(1, 15, 100)`, the next batch can be `(1, 24, 100)`, whatever. That will mean you are only doing batches of size 1 so it will be noisy, but you can easily write something to have different lengths in each batch.
> 2. If you want different lengths within a single batch, you need to use masking. You can signal in-band or out-of-band. In-band would be a special value that indicates that the step should be ignored. Out-of-band would be a separate input with `1` for included steps and `0` for excluded steps. Multiply your loss by the mask, so losses for excluded steps are 0.
> 
> If you do both of the above, each batch will be the length of the longest sequence in that batch. The shorter sequences in that batch will be padded and masked. This should be the most efficient method and the least amount of data each time.
> 
> It would be best to do both. If you use `fit`, everything needs to be padded out to the length of the longest sequence. If there are some really long outliers, that will have a huge impact.
> 
> Here is a somewhat complicated example but includes one way of doing the above. This experiment concatenates input and output, so only uses `max(input len+output len)` instead of `max(input)+max(output)`, but that is a proof-of-concept and not necessary.
> https://github.com/bstriner/keras-seq2seq
> 
> Not the cleanest code, but if you understand how it works then you should be able to do anything you want.
> 
> Cheers

I have tried you approach. Apparently i am using masking as well now with you approach input is defined as 
`model.add(Masking(mask_value=0. , input_shape=(None, number_of_features)))`
As you can see input shape has None as time steps because as you explained the batches are of different length so that we have maximum efficiency while doing padding. But the problem is that Masking is not in effect with this. It doesn't throws any error but one can see in output of predictions. 
Upon further analyzing i came to know about this thing while keras is computing masking it throws this error. 

> Layer masking_1 is not connected, no input to return. 

>  Layer masking_1 has multiple inbound nodes, hence the notion of "layer input mask" is ill-defined. Use `get_input_mask_at(node_index)` instead.

So basically masking is not in effect due to input shape. 
Can you @fchollet @bstriner help me with this. I am using latest version of keras
